### PR TITLE
Make name, title, frontend_title fields on UFGroup consistent with ContributionPage, Group

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------

Make name, title, frontend_title fields on UFGroup consistent with ContributionPage, Group

Before
----------------------------------------
No name field, frontend_title not required

After
----------------------------------------
name field created, all 3 are required but in most cases `writeRecord` should wrangle if one is provided

Technical Details
----------------------------------------
~~upgrade not done at the stage - likely a test-fight ahead~~

Comments
----------------------------------------
This is the last of these entities where frontend_title as added but not required & will resolve https://github.com/civicrm/civicrm-core/pull/23325

### CODE REVIEW

**Comment by mattwire** [comment]
@eileenmcnaughton Needs rebase

**Comment by eileenmcnaughton** [comment]
[Test Result](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/7717/BKPROF=dfl,SUITES=phpunit-api3,label=bknix-tmp/testReport/) (1 failure / -2)
[api_v3_UFGroupTest.testUFGroupCreateWithWrongParams with data set "APIv4"](https://test.civicrm.org/job/CiviCRM-Core-Matrix-PR/7717/BKPROF=dfl,SUITES=phpunit-api3,label=bknix-tmp/testReport/junit/(root)/api_v3_UFGroupTest/testUFGroupCreateWithWrongParams_with_data_set__APIv4_/)

File: CRM/Upgrade/Incremental/php/FiveSeventy.php [hunk 1]
```diff
@@ -28,7 +28,20 @@ class CRM_Upgrade_Incremental_php_FiveSeventy extends CRM_Upgrade_Incremental_Ba
...
   public function upgrade_5_70_alpha1($rev): void {
+    $this->addTask('Add `name` to `civicrm_uf_group', 'addColumn', 'civicrm_uf_group', 'name',
+      "varchar(64) NULL COMMENT 'Form name.'"
+    );
+    // It is important the line above run before the sql, which will populate the fields
+    // before they are made required.
     $this->addTask(ts('Upgrade DB to %1: SQL', [1 => $rev]), 'runSql', $rev);
+
+    $this->addTask('Make civicrm_uf_group.name required', 'alterColumn', 'civicrm_uf_group', 'name',
+      "varchar(255) NOT NULL  COMMENT 'Form name."
```
Comment (colemanw) [inline CRM/Upgrade/Incremental/php/FiveSeventy.php:39]: In the schema xml the length of the field is 64 but this says 255.

Comment (eileenmcnaughton) [reply CRM/Upgrade/Incremental/php/FiveSeventy.php:39]: that's because of copy & paste - it's a feature

**Comment by eileenmcnaughton** [comment]
Upgrade DB to 5.70.alpha1: SQL (CRM_Upgrade_Incremental_php_FiveSeventy::runSql(5.70.alpha1))
Cannot execute WHERE `frontend_title` IS NULL OR `frontend_title_` = '': DB Error: syntax error

**Comment by eileenmcnaughton** [comment]
@seamuslee001 do you have any idea of what is going on here? Only 1 of the upgrade tests is failing now - at first it was all of them when the syntax was wrong  = the field was first added in 4.7 https://github.com/civicrm/civicrm-core/commit/bf48be46bcf15aa2f6dcedb7b319c36e05b4334c#diff-f57caa38fa2742f1674f00d0ab94574dcf441dec7dba3a2e394c47dd75a7bf53

File: CRM/Upgrade/Incremental/sql/5.71.alpha1.mysql.tpl [hunk 2]
```diff
@@ -1 +1,22 @@
 {* file to handle db changes in 5.71.alpha1 during upgrade *}
+
+-- Add name field, make frontend_title required (in conjunction with php function)
+{if $multilingual}
+  {foreach from=$locales item=locale}
+    UPDATE `civicrm_uf_group`
+    SET `frontend_title_{$locale}` = `title_{$locale}`, `name` = CONCAT(`title_{$locale}`, `id`);
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/sql/5.71.alpha1.mysql.tpl:7]: @eileenmcnaughton the problem is at the end here you have a `;` on this line which will stop the WHERE from being included

File: CRM/Upgrade/Incremental/sql/5.71.alpha1.mysql.tpl [hunk 3]
```diff
@@ -1 +1,22 @@
...
+    WHERE `frontend_title_{$locale}` IS NULL OR `frontend_title_{$locale}` = '';
+
+    UPDATE `civicrm_uf_group`
+    SET `name` = CONCAT(`title_{$locale}`, `id`);
```
Comment (seamuslee001) [inline CRM/Upgrade/Incremental/sql/5.71.alpha1.mysql.tpl:11]: same issue as L7

**Comment by eileenmcnaughton** [comment]
Create index civicrm_uf_group.UI_name (CRM_Upgrade_Incremental_php_FiveSeventyOne::addIndex(civicrm_uf_group  ,Array,UI))
Error executing task "Create index civicrm_uf_group.UI_name"

In Error.php line 971:
                                          
  [Civi\Core\Exception\DBQueryException]  
  DB Error: already exists

**Comment by eileenmcnaughton** [comment]
Create index civicrm_uf_group.UI_name (CRM_Upgrade_Incremental_php_FiveSeventyOne::addIndex(civicrm_uf_group,**Array**,UI))

**Comment by eileenmcnaughton** [comment]
@seamuslee001 I'm still missing a trick here & I just can't spot it

Upgrade DB to 5.71.alpha1: SQL (CRM_Upgrade_Incremental_php_FiveSeventyOne::runSql(5.71.alpha1))
Make civicrm_uf_group.name required (CRM_Upgrade_Incremental_php_FiveSeventyOne::alterColumn(civicrm_uf_group,name,varchar(64) NOT NULL  COMMENT 'Form name.'))
Create index civicrm_uf_group.UI_name (CRM_Upgrade_Incremental_php_FiveSeventyOne::addIndex(civicrm_uf_group,name,UI))
Error executing task "Create index civicrm_uf_group.UI_name"

In Error.php line 971:
                                          
  [Civi\Core\Exception\DBQueryException]  
  DB Error: already exists                
                                          

Exception trace:

**Comment by seamuslee001** [comment]
@eileenmcnaughton if you look at the xml in New Line 272 you will see there is already a unique index for name

**Comment by eileenmcnaughton** [comment]
@seamuslee001 ah....

**Comment by eileenmcnaughton** [comment]
@seamuslee001 it passed!

**Comment by eileenmcnaughton** [comment]
@colemanw @seamuslee001 can we merge this now?

**Comment by colemanw** [comment]
Yep let's not let the upgrade code get stale.

**Comment by agileware-justin** [comment]
Good to see this change implemented, will make the front-end integration just a bit easier and more consistent.

**Comment by eileenmcnaughton** [comment]
yeah - it has been a bit of a slog - but we were in a kinda half state & that seemed to keep causing issues to come up

**Pull request merged**